Objective-C/CBLProperties.mm

```diff
@@ -146,97 +145,403 @@ - (BOOL) containsObjectForKey: (NSString*)key {
+                _properties[key] = nil;
+            }
+        } else {
+            if ([value isKindOfClass: [NSArray class]]) {
+                for (id obj in value) {
+                    CBLSubdocument* subdoc = $castIf(CBLSubdocument, obj);
+                    [subdoc invalidate];
+                }
+            }
+            _properties[key] = nil;
+        }
+        
+    }
+    _changedNames = nil;
     _hasChanges = NO;
 }
 
-
-- (FLSharedKeys) sharedKeys { return nullptr; }
+#pragma mark - INTERNAL: CBLJSONEncoding
 
 
-- (void) propertiesDidChange { }
+- (id) encodeAsJSON {
+    return [self currentPropertiesAsJson: YES];
+}
 
 
 #pragma mark - PRIVATE:
 
+- (void) markChanges {
+    self.hasChanges = YES;
+}
 
-- (nullable NSDictionary*) savedProperties {
-    if (_properties && !self.hasChanges)
-        return _properties;
-    if (_root)
-        return [self fleeceValueToObject: (FLValue)_root];
-    else
-        return _rootProps;
+
+- (void) markChangedKey: (NSString*)key {
+    [self markChangedKeys: @[key]];
 }
 
 
-- (void) mutateProperties {
-    if (!_properties)
-        _properties = [self.savedProperties mutableCopy];
+- (void) markChangedKeys: (NSArray*)keys {
+    if (!_changedNames)
+        _changedNames = [NSMutableSet set];
+    
+    for (NSString* key in keys) {
+        if (_properties[key] != nil && ![_changedNames containsObject: key])
+            [_changedNames addObject: key];
+    }
+    [self markChanges];
+}
+
+
+/** Convert properties content into Subdocuments where dictionaries are found.*/
+- (NSMutableDictionary*) convertProperties: (NSDictionary*)properties {
+    NSMutableDictionary* result = [properties mutableCopy];
+    for (NSString* key in properties) {
+        id oldValue = _properties ? _properties[key] : nil;
+        id nuValue = properties[key];
+        result[key] = [self convertValue: nuValue oldValue: oldValue forKey: key];
+    }
+    return result;
+}
+
+
+- (id) convertValue: (id)value oldValue: (nullable id)oldValue forKey: (NSString*)key {
+    if ([value isKindOfClass: [NSDictionary class]]) {
+        CBLSubdocument* subdoc = nil;
+        // If the old value is a subdocument, use the old value:
+        if ([oldValue isKindOfClass: [CBLSubdocument class]]) {
+            subdoc = (CBLSubdocument*)oldValue;
+            subdoc.properties = value;
+        } else
+            subdoc = [self subdocumentWithRoot: nullptr withProperties: value forKey: key];
+        return subdoc;
+    } else if ([value isKindOfClass: [NSArray class]]) {
+        NSArray* oldArray = $castIf(NSArray, oldValue);
+        NSArray* nuArray = (NSArray*)value;
+        NSMutableArray* result = [NSMutableArray arrayWithCapacity: nuArray.count];
+        for (int i = 0; i < nuArray.count; i++) {
+            id oldValue = oldArray && i < oldArray.count ? oldArray[i] : nil;
+            id nuValue = nuArray[i];
+            result[i] = [self convertValue: nuValue oldValue: oldValue forKey: key];
```

Comment:
This assumes the old and new values at the same indexes correspond to each other. But that's not the case if items have been inserted or removed in the middle of the array. Or if the two arrays are entirely different.